aws/s3.go

```diff
@@ -0,0 +1,295 @@
+package aws
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/gruntwork-io/cloud-nuke/logging"
+	"github.com/gruntwork-io/gruntwork-cli/errors"
+)
+
+// getS3BucketRegion returns S3 Bucket region.
+func getS3BucketRegion(svc *s3.S3, bucketName *string) (string, error) {
+	input := &s3.GetBucketLocationInput{
+		Bucket: bucketName,
+	}
+
+	result, err := svc.GetBucketLocation(input)
+	if err != nil {
+		return "", err
+	}
+
+	if result.LocationConstraint == nil {
+		// GetBucketLocation returns nil for us-east-1
+		// https://github.com/aws/aws-sdk-go/issues/1687
+		return "us-east-1", nil
+	}
+	return *result.LocationConstraint, nil
+}
+
+// getS3BucketTags returns S3 Bucket tags.
+func getS3BucketTags(svc *s3.S3, bucketName *string) ([]map[string]string, error) {
+	input := &s3.GetBucketTaggingInput{
+		Bucket: bucketName,
+	}
+
+	tags := []map[string]string{}
+
+	result, err := svc.GetBucketTagging(input)
+	if err != nil {
+		return tags, err
+	}
+
+	for _, tagSet := range result.TagSet {
+		tags = append(tags, map[string]string{"Key": *tagSet.Key, "Value": *tagSet.Value})
+	}
+
+	return tags, err
+}
+
+// getAllS3Buckets lists and returns a map of per region AWS S3 buckets which were created before excludeAfter
+func getAllS3Buckets(session *session.Session, excludeAfter time.Time, bucketNameSubStr string) (map[string][]*string, error) {
+	svc := s3.New(session)
+
+	input := &s3.ListBucketsInput{}
+
+	output, err := svc.ListBuckets(input)
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+
+	var bucketNamesPerRegion = make(map[string][]*string)
+
+OUTER:
```

Comment:
We generally prefer `return` than `continue`, especially if you have a `continue` with a label. Also, it will be easier to read if you get all the buckets in a list, and then massage it to a map as opposed to going straight to map.

For example:

```go
bucketsToInclude := []Bucket{}
for _, bucket := range output.Buckets {
  if shouldIncludeBucket(bucket) {
    bucketsToInclude = append(bucketsToInclude, bucket)
  }
}
for _, bucket := range bucketsToInclude {
  bucketRegion, err := getS3BucketRegion(svc, bucket.Name)
  if err != nil {
    // log some message
  } else {
    // add to bucketNamesPerRegion
  }
}
```

Reply:
Removed the continue OUTER - but building the map on the go as the code is readable after moving filtering code to shouldIncludeBucket.